### PR TITLE
LDAP - Users not created on the fly upon first login fixed

### LINKED ISSUE #539: v2 LDAP - Users not created on the fly upon first login
Quoting myself from #505:

> I've been testing the LDAP and can confirm that the login itself works fine with `passport-ldapauth`. However, I had to create the users internally in the user management page before they could login. I suspect this is not the desired use case and users should be created on first login, as Mailtrain accepts the ldap properties `newUserRole` and `newNamespaceId`, right?

When logging in with a wrong username & password pair, the form gives a `Invalid user or password` message, whereas using a valid user and password combination, an **permission denied error** appears.

### PR DESCRIPTION
**Pull request by Charlo270398**
Resolves #539

### CODE REVIEW

File: server/lib/passport.js [hunk 1]
```diff
@@ -197,14 +197,16 @@ if (LdapStrategy) {
             if (err instanceof interoperableErrors.NotFoundError) {
                 const userId = await users.create(null, {
                     username: profile[config.ldap.uidTag],
+                    email: profile.mail,
                     role: config.ldap.newUserRole,
-                    namespace: config.ldap.newUserNamespaceId
+                    namespace: config.ldap.newUserNamespaceId,
+                    name: profile.cn
```
Comment (bures) [inline server/lib/passport.js]: The same as below - hardwired field names.

File: server/models/users.js [hunk 2]
```diff
@@ -27,7 +27,7 @@ const namespaceHelpers = require('../lib/namespace-helpers');
 
 const allowedKeys = new Set(['username', 'name', 'email', 'password', 'namespace', 'role']);
 const ownAccountAllowedKeys = new Set(['name', 'email', 'password']);
-const allowedKeysExternal = new Set(['username', 'namespace', 'role']);
+const allowedKeysExternal = new Set(['username', 'namespace', 'role', 'email', 'name']);
```
Comment (bures) [inline server/models/users.js]: Why is this? As far as I remember, email and name are taken from LDAP. Thus they shouldn't be allowed to be changed.

File: server/models/users.js [hunk 3]
```diff
@@ -167,9 +167,9 @@ async function _validateAndPreprocess(tx, entity, isCreate, isOwnAccount) {
 async function create(context, user) {
     let id;
     await knex.transaction(async tx => {
-        await shares.enforceEntityPermissionTx(tx, context, 'namespace', user.namespace, 'manageUsers');
 
         if (passport.isAuthMethodLocal) {
+            await shares.enforceEntityPermissionTx(tx, context, 'namespace', user.namespace, 'manageUsers');
```
[the remark this answers is not in the document]
Comment (Charlo270398) [reply server/models/users.js]: When a user who isn't into the BD executes that line, **enforceEntityPermissionTx** function throws a _throwPermissionDenied_ exception and _create_ function stops. It's impossible to find au user from LDAP into the BD when he is logging for first time, so I think this need a solution anyway.

[the remark this answers is not in the document]
Comment (bures) [reply server/models/users.js]: But when the user is created by Mailtrain when logging in via LDAP for the first time, the call to `create` uses getAdminContext and that should grant it the needed permissions. Thus the check could be there for all the paths. Isn't it?

File: server/lib/passport.js [hunk 4]
```diff
@@ -197,14 +197,16 @@ if (LdapStrategy) {
...
                 });
 
                 return {
                     id: userId,
-                    username: profile[config.ldap.uidTag],
-                    name: profile[config.ldap.nameTag],
+                    username: profile.uid,
```
[the remark this answers is not in the document]
Comment (Charlo270398) [reply server/lib/passport.js]: Ok, you're right. I did the pull req before that commit, so I didn't know about mailTag.

File: server/models/users.js (hunk 2 above)
Comment (Charlo270398) [reply server/models/users.js]: That was an implementation change, we wanted to save email and name into BD. But you're right.

File: server/models/users.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Charlo270398) [reply server/models/users.js]: I think is using **null** context: 

imagen

Use getAdminContext should correct the error.

[the remark this answers is not in the document]
Comment (bures) [reply server/models/users.js]: Oh. `null` is an error. In the past it meant the admin context. But then I refactored it to getAdminContext and obviously forgot this piece of code. Yes, please replace it with getAdminContext.

**Comment by Charlo270398** [comment]
48b3411 commit should fix it

File: server/lib/passport.js [hunk 5]
```diff
@@ -195,18 +195,16 @@ if (LdapStrategy) {
...
+                    namespace: config.ldap.newUserNamespaceId
                 });
 
                 return {
                     id: userId,
-                    username: profile.uid,
-                    name: profile.cn,
+                    username: profile[config.ldap.uidTag],
+                    name: profile[config.ldap.nameTag],
                     email: profile.mail,
```
Comment (bures) [inline server/lib/passport.js]: Please, could you also replace this by the tag from config. This was forgotten in one of the previous commits. And the same on line 192. Thanks.

**Comment by Charlo270398** [comment]
Like this?

**Comment by bures** [comment]
Yep. Great. Thanks for testing it. I don't have an LDAP based deployment, thus it's difficult for me to test it.
However, why is the reports CUD part of the PR? Thanks.

**Comment by Charlo270398** [comment]
Because I don't usually use revert commands and I had problems removing wrong commits.

**Comment by bures** [comment]
No problem. Will you remove the change to reports CUD so that I can merge this PR. Alternatively, you can open a new PR and discard this one if it is easier for you. Thanks.

**Comment by Charlo270398** [comment]
I think the last one is easier :sweat_smile:

**Pull request closed without merging**